src/mono/mono/metadata/class-init.c

```diff
@@ -1004,22 +1003,25 @@ mono_class_create_bounded_array (MonoClass *eclass, guint32 rank, gboolean bound
 	GSList *list, *rootlist = NULL;
 	int nsize;
 	char *name;
-	MonoImageSet* image_set;
+	MonoMemoryManager *mm;
 
 	if (rank > 1)
 		/* bounded only matters for one-dimensional arrays */
 		bounded = FALSE;
 
 	image = eclass->image;
-	image_set = class_kind_may_contain_generic_instances ((MonoTypeKind)eclass->class_kind) ? mono_metadata_get_image_set_for_class (eclass) : NULL;
 
+	// FIXME: Optimize this
+	mm = class_kind_may_contain_generic_instances ((MonoTypeKind)eclass->class_kind) ? mono_metadata_get_mem_manager_for_class (eclass) : NULL;
 	/* Check cache */
 	cached = NULL;
 	if (rank == 1 && !bounded) {
-		if (image_set) {
-			mono_image_set_lock (image_set);
-			cached = (MonoClass *)g_hash_table_lookup (image_set->szarray_cache, eclass);
-			mono_image_set_unlock (image_set);
+		if (mm) {
+			mono_mem_manager_lock (mm);
+			if (!mm->szarray_cache)
+				mm->szarray_cache = g_hash_table_new_full (mono_aligned_addr_hash, NULL, NULL, NULL);
```

Comment:
Do these lazily-created hash tables ever get freed? Am I just missing it?

Reply:
Not right now, but right now memory managers are never freed.

Reply:
We do have a free function though - it would be nice to add those ASAP before we forget.